<a href="https://colab.research.google.com/github/PedroH-07/Checkpoint2-SERS/blob/main/Checkpoint2_SERS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Appliance Energy Prediction**

In [ ]:
#Importando Bibliotecas
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
import numpy as np

In [ ]:
# colocando o dataset e pritando as informações dele
df = pd.read_csv('/content/energydata_complete.csv')

print(df.head())
print(df.info())
print(df.describe())

                  date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility

In [ ]:
X = df.drop(['date', 'Appliances'], axis=1)
y = df['Appliances']

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [ ]:
# Regressão Linear
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred_lr = lr.predict(X_test)

In [ ]:
#Árvore de Regressão (Decision Tree)
dt = DecisionTreeRegressor(random_state=42)
dt.fit(X_train, y_train)
y_pred_dt = dt.predict(X_test)

In [ ]:
#Random Forest
rf = RandomForestRegressor(random_state=42)
rf.fit(X_train, y_train)
y_pred_rf = rf.predict(X_test)

In [ ]:
# Avaliação dos modelos usando R², RMSE e MAE
def print_metrics(y_true, y_pred, model_name):
    r2 = r2_score(y_true, y_pred)
    rmse = np.sqrt(mean_squared_error(y_true, y_pred))
    mae = mean_absolute_error(y_true, y_pred)
    print(f"\n{model_name}:")
    print(f"R² = {r2:.4f}")
    print(f"RMSE = {rmse:.2f}")
    print(f"MAE = {mae:.2f}")

In [ ]:
print_metrics(y_test, y_pred_lr, "Regressão Linear")
print_metrics(y_test, y_pred_dt, "Árvore de Decisão")
print_metrics(y_test, y_pred_rf, "Random Forest")


Regressão Linear:
R² = 0.1693
RMSE = 91.17
MAE = 52.55

Árvore de Decisão:
R² = 0.1798
RMSE = 90.60
MAE = 38.87

Random Forest:
R² = 0.5322
RMSE = 68.42
MAE = 32.90


Conclusão:

- **Random Forest** tem o maior valor de R² (0.5322), indicando que explica uma proporção maior da variância no consumo de energia em comparação com a Regressão Linear (0.1693) e a Árvore de Decisão (0.1798).
- **Random Forest** também tem os menores valores de RMSE (68.42) e MAE (32.90), o que significa que suas previsões estão, em média, mais próximas dos valores reais do que os outros dois modelos.

Portanto, o modelo **Random Forest** é o melhor entre os três para explicar e prever o consumo de energia com base nesta análise.

# **Smart Grid Stability**